Update muon shower (HMT) emulator

kakwok opened the pull request:
#### PR description:

This PR updates muon shower (HMT) emulator based on the data/emulator comparison study with Run 3 data.  Most changes are done in collaboration with TMB developer @tahuang1991 .

Changes for CSC trigger primitive emulator:
 - Output a vector of valid HMT bits within TMB readout window
 - Remove dead-time selection for anode hits in anode HMT emulation
 - Add peak-check option for cathode HMT emulation
 - Add dead-time consideration for cathode HMT emulation
 - Add configurable matching logic per-station/ring chamber

Changes for CSC Packer/Unpacker:
 - Assign BX for anode/cathode and matched HMT during unpacking
 - Update corresponding packing

Changes for CSCShowerDigi:
 - Add BX property
 - Add number of wire/comparator hits
   - Can be filled from emulator or from TMB header  

#### PR validation:

Data/Emulation comparison studies with Run 3 data are presented here:
(https://indico.cern.ch/event/1161409/contributions/4965616/attachments/2481743/4260443/HMT_firstLook_L1DPG.pdf)

Unpack-pack-unpack sequence was checked with this config:
https://gist.github.com/tahuang1991/3805853ff6c0b340605afe1f461822c8

tahuang1991 committed:
Rebase step 2

Commit:
code format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31385

- This PR adds an extra 100KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
A new Pull Request was created by  @kakwok  for master.

It involves the following packages:

- DataFormats/CSCDigi (**simulation**)
- EventFilter/CSCRawToDigi (**reconstruction**)
- L1Trigger/CSCTriggerPrimitives (**l1**)


@civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @cecilecaillol can you please review it and eventually sign? Thanks.
@giovanni-mocellin, @barvic, @rovere, @eyigitba, @valuev, @missirol, @Martin-Grunewald, @ptcox this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

civanch commented:
please test

cmsbuild commented:
-1

**Failed Tests**:   Build 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/26610/summary.html
**COMMIT**: af71bd75a4995ce1eed5d248a7aa77db5964696f
**CMSSW**: CMSSW_12_5_X_2022-08-02-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38941/26610/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTSetup.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTTokens.cc
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/GTCollections.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockME.cc: In member function 'virtual bool l1t::stage2::emtf::MEBlockUnpacker::unpack(const l1t::Block&, l1t::UnpackerCollections*)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockME.cc:307:47: error: no matching function for call to 'CSCShowerDigi::CSCShowerDigi(int, int, CSCDetId)'
  307 |                               Hit_.CSC_DetId());
      |                                               ^
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/DataFormats/CSCDigi/interface/CSCShowerDigiCollection.h:5,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFCollections.h:15,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_12_5_X_2022-08-02-1100/src/EventFilter/L1TRawToDigi/plugins/implementations_stage2/EMTFBlockME.cc:5:
</pre>

tahuang1991 committed:
Merged HMT_emulation_12_5_X from repository kakwok with cms-merge-topic

tahuang1991 committed:
1.update CSCShowerDigi constructor and the codes that used the constructor
2.Add two more shower types: EMTFShower and GMTShower
3.removed comment-out lines in CSCAnodeLCTProcessor.cc

kakwok committed:
Merge pull request #1 from tahuang1991/HMT_emulation_12_5_X

Update the CSCDigi shower constructor and also where the constructor is called

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31422

- This PR adds an extra 72KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31422/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31422/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

tahuang1991 committed:
fixed the code-format in CSCRawToDigi from scram build code-format

tahuang1991 committed:
Merge pull request #2 from tahuang1991/HMT_emulation_12_5_X

fixed the code-format in CSCRawToDigi from scram build code-format

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31423

- This PR adds an extra 76KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @cecilecaillol can you please check and sign again.

kakwok commented:
@civanch We have fixed the remaining places where `CSCShowerDigi` is constructed. Can you please start the test again?

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/26652/summary.html
**COMMIT**: b6a117e394aa98dcf0afa1e60f98d0abf113c2bf
**CMSSW**: CMSSW_12_5_X_2022-08-05-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38941/26652/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3691612
 - DQMHistoTests: Total failures: 56
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 3691533
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.004 KiB( 50 files compared)
 - DQMHistoSizes: changed ( 312.0 ): 0.004 KiB        MessageLogger/Warnings
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-05-1100+b7c06f/51988/triggerResults): no differences found

cecilecaillol commented:
+l1

civanch commented:
+1

clacaputo commented on DataFormats/CSCDigi/src/classes_def.xml:
```diff
@@ -39,7 +39,9 @@
    </class>
-   <class name="CSCShowerDigi" ClassVersion="11">
+   <class name="CSCShowerDigi" ClassVersion="13">
+    <version ClassVersion="13" checksum="2873709471"/>
```
Hi @kakwok , why do you need to increase by two the versioning?

tahuang1991 replied:
we made the some major changes twice on CSC shower digi data format during the internal code development and  debug.  And the classversion and checksum are taken from compile suggestions

kakwok replied:
@tahuang1991 I think we can just use 12 and the checksum, we don't need to skip version 12. 
@clacaputo I'm preparing the update.

Commit:
change class version to 12

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31466

- This PR adds an extra 292KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @cecilecaillol can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/26719/summary.html
**COMMIT**: 35ff4e1f120b45366ae4db4ecb5112935336eb8d
**CMSSW**: CMSSW_12_5_X_2022-08-08-2300/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/38941/26719/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3691612
 - DQMHistoTests: Total failures: 39
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3691551
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-08-2300+b7c06f/52043/triggerResults): no differences found

civanch commented:
+1

cecilecaillol commented:
+l1

jpata commented on DataFormats/CSCDigi/src/CSCShowerDigi.cc:
```diff
@@ -35,6 +53,42 @@ bool CSCShowerDigi::isNominalOutOfTime() const { return bitsOutOfTime() >= kNomi
+      showerStr = "UnknownShowerType";
+  }
+
+  //return o << (showerType == 0
+  //                 ? "Invalid ShowerType"
+  //                 : (showerType == 1 ? "ALCTShower" : (showerType == 2 ? "CLCTShower" : "correlatedLCTShower")))
```
commented-out code should be removed

kakwok replied:
@jpata Done in [5eb6603](https://github.com/cms-sw/cmssw/pull/38941/commits/5eb66037a3fe46e47876ae64512ddcb10a5cdc49)

Commit:
remove commented lines

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31634

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @mandrenguyen, @cecilecaillol can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/26911/summary.html
**COMMIT**: 5eb66037a3fe46e47876ae64512ddcb10a5cdc49
**CMSSW**: CMSSW_12_5_X_2022-08-18-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38941/26911/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 3693040
 - DQMHistoTests: Total failures: 45
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3692973
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 212 log files, 49 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_5_X_2022-08-18-1100+b7c06f/52204/triggerResults): no differences found

civanch commented:
+1

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:
```diff
@@ -1276,8 +1273,33 @@ CSCALCTDigi CSCAnodeLCTProcessor::getBestALCT(int bx) const { return bestALCT[bx
 /** Returns shower bits */
-CSCShowerDigi CSCAnodeLCTProcessor::readoutShower() const { return shower_; }
+std::vector<CSCShowerDigi> CSCAnodeLCTProcessor::readoutShower() const {
+  unsigned minbx_readout = CSCConstants::LCT_CENTRAL_BX - l1a_window_width / 2;
```
Would a check that `CSCConstants::LCT_CENTRAL_BX ->= l1a_window_width / 2` be worth (in the constructor)?

It is true that if `minbx_readout` overflows, the loop at [L1289](https://github.com/cms-sw/cmssw/pull/38941/files#diff-bbb72f657aa720b2f0097ffdd8af20ba5d3940490879aae6c19c2f18105279c6R1289) does not even start, and therefore it is harmless anyhow

kakwok replied:
moved to `minbx_readout` and `maxbx_readout` to header files and added `assert` in the constructor 712e670

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:
```diff
@@ -1201,52 +1199,48 @@ CSCCLCTDigi CSCCathodeLCTProcessor::getSecondCLCT(int bx) const {
 /** Returns shower bits */
-CSCShowerDigi CSCCathodeLCTProcessor::readoutShower() const { return shower_; }
+std::vector<CSCShowerDigi> CSCCathodeLCTProcessor::readoutShower() const {
+  unsigned minbx_readout = CSCConstants::LCT_CENTRAL_BX - tmb_l1a_window_size / 2;
```
Same as above...

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:
```diff
@@ -1256,13 +1250,51 @@ void CSCCathodeLCTProcessor::encodeHighMultiplicityBits(
+          (hitsInTime[minbx] == hitsInTime[maxbx + 1] and hitsInTime[bx] < hitsInTime[bx + 1]))
+        isPeak = false;  //next bx would have more hits or in the center
+    }
+    if (dead_count > 0) {
+      dead_count--;
+      dead_status = true;
+    } else
+      dead_status = false;
```
`dead_status` could be a local variable in the scope:
```suggestion
    bool dead_status = dead_count > 0:
    if (dead_status)
      dead_count--;
```

kakwok replied:
Fixed as suggested

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCCathodeLCTProcessor.cc:
```diff
@@ -1256,13 +1250,51 @@ void CSCCathodeLCTProcessor::encodeHighMultiplicityBits(
+    // do nothing if there are not enough layers with hits
+    if (layersWithHits[bx].size() >= minLayersCentralTBin_ and !dead_status and isPeak) {
+      // assign the bits
+      for (unsigned i = 0; i < station_thresholds.size(); i++) {
+        if (this_hitsInTime >= station_thresholds[i]) {
+          this_inTimeHMT = i + 1;
+          dead_count = deadtime;
+        }
+      }
```
If this has to be overwritten at every station_threshold, it is probably more convenient:
```suggestion
      if (station_thresholds.size() > 0) {
        for (unsigned i = station_thresholds.size() - 1; i >= 0; i--) {
          if (this_hitsInTime >= station_thresholds[i]) {
            this_inTimeHMT = i + 1;
            dead_count = deadtime;
            break;
          }
        }
      }
```

kakwok replied:
Hi @perrotta 
`this_inTimeHMT` needs to count how many `station_thresholds`(loose, nominal, tight) it passes. It is used to set the bit for the `CSCShowerDigi` object [here](https://github.com/cms-sw/cmssw/pull/38941/files#diff-fa43c5c8243291071f6787aa0b1586f9244ece75c3f67547023b0e81ab7431caR12)

So we can't have the `break` and the original implementation seems more natural to me.

I'm also addressing the other comments.

perrotta replied:
@kakwok but here you are not counting how many station_thresholds it passes, just the index +1 of the last station over threshold. Which is the first one encountered if you loop in the reverse order, as I suggested...

kakwok replied:
@perrotta I see what you mean now. I'm fine with counting in reverse order then.

kakwok replied:
fixed as suggested.

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCMotherboard.cc:
```diff
@@ -365,7 +369,26 @@ std::vector<CSCCorrelatedLCTDigi> CSCMotherboard::readoutLCTs() const {
 
-CSCShowerDigi CSCMotherboard::readoutShower() const { return shower_; }
+std::vector<CSCShowerDigi> CSCMotherboard::readoutShower() const {
+  unsigned minbx_readout = CSCConstants::LCT_CENTRAL_BX - tmb_l1a_window_size / 2;
```
Same as above

perrotta commented on EventFilter/CSCRawToDigi/src/CSCTMBHeader2020_CCLUT.cc:
```diff
@@ -161,20 +161,42 @@ std::vector<CSCCorrelatedLCTDigi> CSCTMBHeader2020_CCLUT::CorrelatedLCTDigis(uin
+  //L1A_TMB_WINDOW is not included in below formula
+  //correct version:  CSCConstants::LCT_CENTRAL_BX - bits.pop_l1a_match_win + L1A_TMB_WINDOW/2;
+  // same for anode HMT and cathode HMT
+  uint16_t bx = CSCConstants::LCT_CENTRAL_BX - bits.pop_l1a_match_win;
```
There are quite several dangerous differences between unsigned integers in thus code: I assume they are all fully under control...

perrotta commented on EventFilter/CSCRawToDigi/src/CSCTMBHeader2020_CCLUT.cc:
```diff
@@ -272,18 +294,47 @@ void CSCTMBHeader2020_CCLUT::addCorrelatedLCT1(const CSCCorrelatedLCTDigi& digi)
 
 void CSCTMBHeader2020_CCLUT::addShower(const CSCShowerDigi& digi) {
   uint16_t hmt_bits = (digi.bitsInTime() & 0x3) + ((digi.bitsOutOfTime() & 0x3) << 2);
+  //not valid LCT shower, then in-time bits must be 0. keep out-of-time HMT
+  if (not digi.isValid())
+    hmt_bits = ((digi.bitsOutOfTime() & 0x3) << 2);
```
Some tiny perf improvement could be achieved with
```suggestion
  uint16_t hmt_bits = digi.isValid()
                      ? (digi.bitsInTime() & 0x3) + ((digi.bitsOutOfTime() & 0x3) << 2)
  //if not valid LCT shower, then in-time bits must be 0. keep out-of-time HMT:
                      : ((digi.bitsOutOfTime() & 0x3) << 2);
```
(and the same in the several occurences of similar patterns in the other parts of this PR)

kakwok replied:
fixed as suggested

perrotta reviewed (commented):
A few sparse comments after looking at the code (quite minor indeed, but you can take them into account in case further updates are needed anyhow)

cecilecaillol commented:
+l1

jpata commented:
+reconstruction
- on the reco side, there is no impact (I understand it's for emulator studies)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @qliphy, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
@kakwok please let us know when do you plan to implement the [updates](https://github.com/cms-sw/cmssw/pull/38941#pullrequestreview-1078520188) that you agreed to apply several days ago already

kakwok commented:
Hi @perrotta , I am working on it now. Will finish today

Commit:
move minbx_readout, maxbx_readout to header, assert minbx_readout>0

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31853

- This PR adds an extra 332KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @mandrenguyen, @cecilecaillol can you please check and sign again.

perrotta commented:
Thank you @kakwok 
As we discussed privately, did you check whether the later modifications of `l1a_window_width`
e.g. in https://github.com/cms-sw/cmssw/blob/712e67041262b5428b5f7101256cf2bbdbbf9f04/L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc#L114 should also trigger a modification of `minbx_readout_` and `maxbx_readout_`?

In that example, I think that it the methids `CSCAnodeLCTProcessor::setConfigParameters(...)` (and `CSCAnodeLCTProcessor::checkConfigParameters()`) are needed, `minbx_readout_` and `maxbx_readout_` should also get updated therein; if those two methods are not needed, insted, they could simply get removed.

The same could happen in the other classes: please check

Commit:
update readout range config

kakwok commented:
@perrotta Thanks!
I updated the `minbx_readout_` and `maxbx_readout` values in the functions. 
I checked that the two variables should have final values the end of `setConfigParameters`, so I moved the `assert` there.

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @mandrenguyen, @cecilecaillol can you please check and sign again.

jpata commented:
@cmsbuild please test

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:
```diff
@@ -135,6 +134,9 @@ void CSCAnodeLCTProcessor::setConfigParameters(const CSCDBL1TPParameters* conf)
   }
+  minbx_readout_ = CSCConstants::LCT_CENTRAL_BX - l1a_window_width / 2;
+  maxbx_readout_ = CSCConstants::LCT_CENTRAL_BX + l1a_window_width / 2;
+  assert(minbx_readout_ > 0);
```
This assert does not work: `minbx_readout_` is unsigned, and it is above 0 by construction
```suggestion
  assert(l1a_window_width / 2 <= CSCConstants::LCT_CENTRAL_BX);
```

perrotta replied:
Moreover: I don't see this setConfigParameters called anywhere.
Who calls it?
Perhaps, another assert also in the constructor won't hurt

kakwok replied:
It's called by the CSCTriggerPrimitives builders [here](https://github.com/cms-sw/cmssw/blob/master/L1Trigger/CSCTriggerPrimitives/plugins/CSCTriggerPrimitivesProducer.cc#L243) and [here](https://github.com/cms-sw/cmssw/blob/master/L1Trigger/CSCTriggerPrimitives/src/CSCTriggerPrimitivesBuilder.cc#L90).

perrotta commented on L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:
```diff
@@ -171,15 +173,18 @@ void CSCAnodeLCTProcessor::checkConfigParameters() {
   CSCBaseboard::checkConfigParameters(accel_mode, max_accel_mode, def_accel_mode, "accel_mode");
   CSCBaseboard::checkConfigParameters(l1a_window_width, max_l1a_window_width, def_l1a_window_width, "l1a_window_width");
+
+  minbx_readout_ = CSCConstants::LCT_CENTRAL_BX - l1a_window_width / 2;
```
These are already set in the calling method `setConfigParameters`, there is no reason to redefine them here.
Perhaps it could make sense to move the `assert` here (it is a "check" in fact), provided the `setConfigParameters` is called every time after (or inside...) the constructor.

kakwok replied:
Yes, `checkConfigParameters` is only called in the constructor or within `setConfigParameters`: [here](https://cmssdt.cern.ch/lxr/search?%21v=CMSSW_12_6_X_2022-08-28-2300&_filestring=&_string=checkConfigParameters)
Yes, I think it's sufficient to keep the assert in `checkConfigParameters`, since it's already called in the constructor.

Commit:
fix assert

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31867

- This PR adds an extra 360KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @mandrenguyen, @cecilecaillol can you please check and sign again.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/summary.html
**COMMIT**: 82c8581965d0c6ef872b3217e88f57603bf20c78
**CMSSW**: CMSSW_12_6_X_2022-08-29-1100/el8_amd64_gcc10
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/38941/27168/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [7.3](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrix-results/7.3_CosmicsSPLoose_UP18+CosmicsSPLoose_UP18+DIGICOS_UP18+RECOCOS_UP18+ALCACOS_UP18+HARVESTCOS_UP18)```7.3_CosmicsSPLoose_UP18+CosmicsSPLoose_UP18+DIGICOS_UP18+RECOCOS_UP18+ALCACOS_UP18+HARVESTCOS_UP18/step3_CosmicsSPLoose_UP18+CosmicsSPLoose_UP18+DIGICOS_UP18+RECOCOS_UP18+ALCACOS_UP18+HARVESTCOS_UP18.log```
- [2018.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrix-results/2018.1_TTbar_13_UP18+TTbarFS_13_UP18+HARVESTUP18FS+MINIAODMCUP18FS)```2018.1_TTbar_13_UP18+TTbarFS_13_UP18+HARVESTUP18FS+MINIAODMCUP18FS/step1_TTbar_13_UP18+TTbarFS_13_UP18+HARVESTUP18FS+MINIAODMCUP18FS.log```
- [139.001](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrix-results/139.001_RunMinimumBias2021+RunMinimumBias2021+HLTDR3_2021+RECODR3_MinBiasOffline+HARVESTD2021MB)```139.001_RunMinimumBias2021+RunMinimumBias2021+HLTDR3_2021+RECODR3_MinBiasOffline+HARVESTD2021MB/step3_RunMinimumBias2021+RunMinimumBias2021+HLTDR3_2021+RECODR3_MinBiasOffline+HARVESTD2021MB.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [1330.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15)
- [13234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrix-results/13234.0_TTbar_14TeV+2021FS+TTbar_14TeV_TuneCP5_Gen+FastSimRun3+HARVESTFastRun3)
</details>



## RelVals-INPUT

- [134.813](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2)```134.813_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2/step2_RunCosmics2015C+RunCosmics2015C+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2.log```
- [136.727](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.727_RunDoubleEGPrpt2016B+RunDoubleEGPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2)```136.727_RunDoubleEGPrpt2016B+RunDoubleEGPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2/step2_RunDoubleEGPrpt2016B+RunDoubleEGPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2.log```
- [136.728](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.728_RunSingleMuPrpt2016B+RunSingleMuPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2)```136.728_RunSingleMuPrpt2016B+RunSingleMuPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2/step2_RunSingleMuPrpt2016B+RunSingleMuPrpt2016B+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [136.736](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.736_RunHLTPhy2016C+RunHLTPhy2016C+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2)
- [136.737](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.737_RunDoubleEG2016C+RunDoubleEG2016C+HLTDR2_2016+RECODR2_2016reHLT_skimDoubleEG_HIPM+HARVESTDR2_skimDoubleEG)
- [136.733](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.733_RunCosmics2016B+RunCosmics2016B+RECOCOSDRUN2+ALCACOSDRUN2+HARVESTDCRUN2)
- [136.746](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.746_RunMuOnia2016C+RunMuOnia2016C+HLTDR2_2016+RECODR2_2016reHLT_skimMuOnia_HIPM+HARVESTDR2_skimMuOnia)
- [136.747](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.747_RunHLTPhy2016D+RunHLTPhy2016D+HLTDR2_2016+RECODR2_2016reHLT_HIPM+HARVESTDR2)
- [136.751](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.751_RunMET2016D+RunMET2016D+HLTDR2_2016+RECODR2_2016reHLT_skimMET_HIPM+HARVESTDR2_skimMET)
- [136.756](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.756_RunZeroBias2016D+RunZeroBias2016D+HLTDR2_2016+RECODR2_2016reHLT_ZB_HIPM+HARVESTDR2ZB)
- [136.762](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.762_RunMET2016E+RunMET2016E+HLTDR2_2016+RECODR2_2016reHLT_skimMET_HIPM+HARVESTDR2_skimMET)
- [136.767](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.767_RunZeroBias2016E+RunZeroBias2016E+HLTDR2_2016+RECODR2_2016reHLT_ZB_HIPM+HARVESTDR2ZB)
- [136.808](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.808_RunMET2017D+RunMET2017D+HLTDR2_2017+RECODR2_2017reHLT_skimMET_Prompt+HARVEST2017_skimMET)
- [136.837](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.837_RunZeroBias2017F+RunZeroBias2017F+HLTDR2_2017+RECODR2_2017reHLT_ZBPrompt+HARVEST2017ZB)
- [136.845](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.845_RunCharmonium2017C+RunCharmonium2017C+HLTDR2_2017+RECODR2_2017reHLT_skimCharmonium_Prompt+HARVEST2017_skimCharmonium)
- [136.864](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.864_RunJetHT2018B+RunJetHT2018B+HLTDR2_2018+RECODR2_2018reHLT_skimJetHT_Offline+HARVEST2018_skimJetHT)
- [136.875](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.875_RunDoubleMuon2018C+RunDoubleMuon2018C+HLTDR2_2018+RECODR2_2018reHLT_Offline+HARVEST2018)
- [136.897](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.897_RunCosmics2021CRUZET+RunCosmics2021CRUZET+RECOCOSDRUN3+ALCACOSDRUN3+HARVESTDCR3)
- [136.899](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/136.899_RunCosmics2021CRAFT+RunCosmics2021CRAFT+RECOCOSDRUN3+ALCACOSDRUN3+HARVESTDCR3)
- [138.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/138.1_PromptCosmics+RunCosmics2021+RECOCOSDPROMPTRUN3+ALCACOSDPROMPTRUN3+HARVESTDCPROMPTRUN3)
- [138.2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/138.2_ExpressCosmics+RunCosmics2021+RECOCOSDEXPRUN3+ALCACOSDEXPRUN3+HARVESTDCEXPRUN3)
- [139.004](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/139.004_RunNoBPTX2021+RunNoBPTX2021+HLTDR3_2021+RECODR3_AlCaTkCosmics_Offline+HARVESTDR3)
- [158.1](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/158.1_QCD_Pt_80_120_13_HI+QCD_Pt_80_120_13_HIINPUT+DIGIHI2018PPRECO+RECOHI2018PPRECO+HARVESTHI2018PPRECO)
- [50203.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/50203.0_H125GGgluonfusion_13+H125GGgluonfusion_13INPUT+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50)
- [50208.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/50208.0_SMS-T1tttt_mGl-1500_mLSP-100_13+SMS-T1tttt_mGl-1500_mLSP-100_13INPUT+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50)
- [1001.2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/1001.2_RunZeroBias2017F+RunZeroBias2017F+TIER0EXPRUN2+ALCAEXPRUN2+ALCAHARVDSIPIXELCAL+ALCAHARVDSIPIXELCALLA+ALCAHARVD4+ALCAHARVDSIPIXELALIHG)
- [11834.13](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/11834.13_TTbar_14TeV+2021PU_mlpf+TTbar_14TeV_TuneCP5_GenSimINPUT+DigiPU+RecoNanoPU+HARVESTNanoPU)
- [12034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/12034.0_TTbar_14TeV+2021Design+TTbar_14TeV_TuneCP5_GenSimINPUT+Digi+RecoNano+HARVESTNano)
- [12434.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/12434.0_TTbar_14TeV+2023+TTbar_14TeV_TuneCP5_GenSimINPUT+Digi+RecoNano+HARVESTNano+ALCA)
- [11846.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/11846.0_ZEE_14+2021PU+ZEE_14TeV_TuneCP5_GenSimINPUT+DigiPU+RecoNanoPU+HARVESTNanoPU)
- [12634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/12634.0_TTbar_14TeV+2023PU+TTbar_14TeV_TuneCP5_GenSimINPUT+DigiPU+RecoNanoPU+HARVESTNanoPU)
- [23234.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/23234.21_TTbar_14TeV+2026D49_ProdLike+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+MiniAOD)
- [39434.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/39434.21_TTbar_14TeV+2026D88_ProdLike+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+MiniAOD)
- [39434.501](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/39434.501_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14INPUT+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [39496.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/39496.0_CloseByPGun_CE_E_Front_120um+2026D88+CE_E_Front_120um_GenSimHLBeamSpotHGCALCloseByINPUT+DigiTrigger+RecoGlobal+HARVESTGlobal)
- [250200.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250200.0_ZEE_13+ZEE_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250202.0_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250200.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250200.17_ZEE_13UP17+ZEE_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250202.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250202.17_TTbar_13UP17+TTbar_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250202.172](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250202.172_TTbar_13UP17+TTbar_13UP17INPUT+DIGIPRMXUP17_PU25_RD+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250202.2](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250202.2_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250202.4](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250202.4_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15APVSimu_PU25+RECOPRMXUP15_HIPM_PU25+HARVESTUP15_PU25)
- [250203.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250203.0_H125GGgluonfusion_13+H125GGgluonfusion_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250203.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250203.17_H125GGgluonfusion_13UP17+H125GGgluonfusion_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250204.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250204.0_QQH1352T_13+QQH1352T_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250204.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250204.17_QQH1352T_13UP17+QQH1352T_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250205.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250205.0_ZTT_13+ZTT_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250205.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250205.17_ZTT_13UP17+ZTT_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250206.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250206.0_ZMM_13+ZMM_13INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [250206.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250206.17_ZMM_13UP17+ZMM_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250207.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250207.0_NuGun_UP15+NuGun_UP15INPUT+DIGIPRMXUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25)
- [500200.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500200.0_ZEE_13+ZEE_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [500202.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500202.0_TTbar_13+TTbar_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [250207.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250207.17_NuGun_UP17+NuGun_UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [250208.17](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/250208.17_SMS-T1tttt_mGl-1500_mLSP-100_13UP17+SMS-T1tttt_mGl-1500_mLSP-100_13UP17INPUT+DIGIPRMXUP17_PU25+RECOPRMXUP17_PU25+HARVESTUP17_PU25)
- [500203.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500203.0_H125GGgluonfusion_13+H125GGgluonfusion_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [500205.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500205.0_ZTT_13+ZTT_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [500204.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500204.0_QQH1352T_13+QQH1352T_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [500206.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500206.0_ZMM_13+ZMM_13INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
- [500207.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b7c06f/27168/runTheMatrixINPUT-results/500207.0_NuGun_UP15+NuGun_UP15INPUT+DIGIPRMXUP15_PU50+RECOPRMXUP15_PU50+HARVESTUP15_PU50)
</details>

kakwok commented:
I can't pin down the exact problem by looking at the log. Will try to reproduce it in a local set-up.

Commit:
set readout window in constructor

kakwok commented:
It seems that the `CSCMotherboard` objects can be created without calling `setConfigParameters`. 
Now [e27b5f3](https://github.com/cms-sw/cmssw/pull/38941/commits/e27b5f3daeceeb6a7ba33893c959c70feaf0a669) sets the values in the constructor. 
Tested in local set-up this fixes the crash.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-38941/31885

- This PR adds an extra 356KB to repository

- Found files with invalid states:
  - DataFormats/CSCDigi/src/classes_def.xml.generated:
    - Added: a5a6c3b12ff3a1ce875efd72286b293c3fbc6424
    - Deleted: af71bd75a4995ce1eed5d248a7aa77db5964696f

cmsbuild commented:
Pull request #38941 was updated. @civanch, @epalencia, @clacaputo, @cmsbuild, @rekovic, @mdhildreth, @jpata, @mandrenguyen, @cecilecaillol can you please check and sign again.

jpata commented:
@cmsbuild please test

tahuang1991 commented:
> It seems that the `CSCMotherboard` objects can be created without calling `setConfigParameters`. Now [e27b5f3](https://github.com/cms-sw/cmssw/pull/38941/commits/e27b5f3daeceeb6a7ba33893c959c70feaf0a669) sets the values in the constructor. Tested in local set-up this fixes the crash.

setConfigParameters is used to read in OTMB/TMB configuration from database and usually it is not used.   The configuration parameter is parsed from python configuration in almost all cases.   So yes,  parameters for HMT should either go to constructor or go to configuration parameters

kakwok commented:
@jpata @perrotta All checks has been passed, and the two back-ports PR (#39260, #39228) are ready as well.

cecilecaillol commented:
+l1

jpata commented:
+reconstruction
- changes since the last reco sig at https://github.com/cms-sw/cmssw/pull/38941#issuecomment-1224020042 were not related to reco
- there are no reco diffs in tests, while some differences in L1 and HLT in the DQM

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

rappoccio commented:
+1
